CppSamples/EditData/SnapGeometryEditsWithRules/SnapGeometryEditsWithRules.cpp

```diff
@@ -0,0 +1,559 @@
+  if (!junctionTable)
+  {
+    return makeExceptionalFutureHelper<std::pair<SubtypeFeatureLayer*, SubtypeFeatureLayer*>>("Unable to create PipelineJunction feature table");
+  }
+  auto junctionLayer = new SubtypeFeatureLayer(junctionTable, this);
+  m_map->operationalLayers()->append(junctionLayer);
+  layerLoadingFutures.emplace_back(load(junctionLayer));
+
+  return QtFuture::whenAll(layerLoadingFutures.begin(), layerLoadingFutures.end()).then(
+  [pipeLayer, deviceLayer](const QList<QFuture<void>>&)
+  {
+    return std::make_pair(pipeLayer, deviceLayer);
+  });
+}
+
+QFuture<void> SnapGeometryEditsWithRules::loadUtilityNetwork()
+{
+  // Add the utility network to the map and load it.
+  auto* utilityNetwork = m_geodatabase->utilityNetworks().constFirst();
+  if (!utilityNetwork)
+  {
+    return makeExceptionalFutureHelper<void>("Unable to create UtilityNetwork");
+  }
+
+  m_map->utilityNetworks()->append(utilityNetwork);
+
+  return load(utilityNetwork);
+}
+
+Esri::ArcGISRuntime::ArcGISFeature* SnapGeometryEditsWithRules::getFeatureFromResult(const QList<Esri::ArcGISRuntime::IdentifyLayerResult*>& identifyResult)
+{
+  if (identifyResult.isEmpty())
+  {
+    return nullptr;
+  }
+
+  QList<IdentifyLayerResult*> sublayerResults = identifyResult.first()->sublayerResults();
+  if (sublayerResults.isEmpty())
+  {
+    return nullptr;
+  }
+
+  QList<GeoElement*> geoElements = sublayerResults.first()->geoElements();
+  if (geoElements.isEmpty())
+  {
+    return nullptr;
+  }
+
+  auto* feature = dynamic_cast<ArcGISFeature*>(geoElements.first());
+  if (!feature)
+  {
+    return nullptr;
+  }
+
+  return feature;
+}
+
+void SnapGeometryEditsWithRules::onMapViewClicked(const QMouseEvent& mouseEvent)
+{
+  if (m_mapView->geometryEditor()->isStarted())
+  {
+    return;
+  }
+
+  // Identify the feature at the tapped location.
+  m_mapView->identifyLayersAsync(mouseEvent.position(), 5, false).then(
+  [this](const QList<IdentifyLayerResult*>& identifyResult)
+  {
+    ArcGISFeature* selectedFeature = getFeatureFromResult(identifyResult);
+
+    // In this sample we only allow selection of point features.
+    // If the identified feature is null or the feature is not a point feature then reset the selection and return.
+    if (!selectedFeature || !selectedFeature->featureTable() || selectedFeature->featureTable()->geometryType() != GeometryType::Point)
+    {
+      resetSelections();
+      return;
+    }
+
+    if (m_selectedFeature != nullptr && selectedFeature != m_selectedFeature)
+    {
+      // If a feature is already selected and the tapped feature is not the selected feature then clear the previous selection.
+      dynamic_cast<FeatureLayer*>(m_selectedFeature->featureTable()->layer())->clearSelection();
+    }
+
+    // Update the selected feature.
+    m_selectedFeature = selectedFeature;
+
+    // Select the feature on the layer.
+    if (auto* layer = dynamic_cast<FeatureLayer*>(m_selectedFeature->featureTable()->layer()))
```

Comment:
```suggestion
    if (auto* layer = static_cast<FeatureLayer*>(m_selectedFeature->featureTable()->layer()); layer)
```
dynamic_cast is not useful if the result is not checked. Looks like you know the cast will work, so go with static_cast.